providers/apis/dydx/switch_over_fetcher.go

```diff
@@ -0,0 +1,153 @@
+package dydx
+
+import (
+	"context"
+	"fmt"
+
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+
+	"github.com/skip-mev/slinky/oracle/config"
+	slinkygrpc "github.com/skip-mev/slinky/pkg/grpc"
+	"github.com/skip-mev/slinky/providers/apis/marketmap"
+	apihandlers "github.com/skip-mev/slinky/providers/base/api/handlers"
+	"github.com/skip-mev/slinky/providers/base/api/metrics"
+	mmclient "github.com/skip-mev/slinky/service/clients/marketmap/types"
+)
+
+var _ mmclient.MarketMapFetcher = &SwitchOverFetcher{}
+
+// SwitchOverFetcher is an implementation of a RestAPIFetcher that wraps a
+// dydx x/prices market map fetcher and a x/marketmap fetcher. The fetcher
+// operates by first fetching the market map from the x/prices API and then
+// fetching the market map from the x/marketmap API. The fetcher will switch
+// over to the x/marketmap API the first time it receives a non-nil market map
+// from the x/marketmap API.
+type SwitchOverFetcher struct {
+	logger *zap.Logger
+
+	// dydxPricesFetcher is the fetcher for the dydx x/prices market map.
+	pricesFetcher mmclient.MarketMapFetcher
+	// marketmapFetcher is the fetcher for the x/marketmap market map.
+	marketmapFetcher mmclient.MarketMapFetcher
+	// switched is true if the fetcher has switched over to the x/marketmap API.
+	switched bool
+}
+
+// DefaultSwitchOverMarketMapFetcher returns a new SwitchOverProvider with the default
+// dYdX x/prices and x/marketmap fetchers.
+func DefaultSwitchOverMarketMapFetcher(
+	logger *zap.Logger,
+	api config.APIConfig,
+	rh apihandlers.RequestHandler,
+	metrics metrics.APIMetrics,
+) (mmclient.MarketMapFetcher, error) {
+	if logger == nil {
+		return nil, fmt.Errorf("logger is nil")
+	}
+	if api.Name != SwitchOverAPIHandlerName {
+		return nil, fmt.Errorf("expected api name %s, got %s", SwitchOverAPIHandlerName, api.Name)
+	}
+	if len(api.Endpoints) != 2 {
+		return nil, fmt.Errorf(
+			"expected two endpoints, got %d; the first end point must be the rest port second endpoint gRPC",
```

Comment:
```suggestion
			"expected two endpoints, got %d; the first endpoint must be the rest port second endpoint gRPC",
```
is it correct the first endpoint must be a rest port? little confused on what they're supposed to be by this error

Reply:
yea supposed to be rest and gRPC

Reply:
ill just remove that from here for now and put it in docs instead